src/document/crdt/tree.ts

```diff
@@ -584,94 +736,96 @@ export class CRDTTree extends CRDTGCElement {
     editedAt: TimeTicket,
   ): Array<TreeChange> {
     // 01. split text nodes at the given range if needed.
-    const [toPos, toRight] = this.findTreePosWithSplitText(range[1], editedAt);
-    const [fromPos, fromRight] = this.findTreePosWithSplitText(
+    const [fromParent, fromLeft] = this.findNodesAndSplitText(
       range[0],
       editedAt,
     );
+    const [toParent, toLeft] = this.findNodesAndSplitText(range[1], editedAt);
 
     // TODO(hackerwins): If concurrent deletion happens, we need to seperate the
     // range(from, to) into multiple ranges.
     const changes: Array<TreeChange> = [];
     changes.push({
       type: TreeChangeType.Content,
-      from: this.toIndex(range[0]),
-      to: this.toIndex(range[1]),
-      fromPath: this.indexTree.treePosToPath(fromPos),
-      toPath: this.indexTree.treePosToPath(toPos),
+      from: this.toIndex(fromParent, fromLeft),
+      to: this.toIndex(toParent, toLeft),
+      fromPath: this.toPath(fromParent, fromLeft),
+      toPath: this.toPath(toParent, toLeft),
       actor: editedAt.getActorID()!,
       value: contents?.length
         ? contents.map((content) => toJSON(content))
         : undefined,
     });
 
     const toBeRemoveds: Array<CRDTTreeNode> = [];
-    // 02. remove the nodes and update linked list and index tree.
-    if (fromRight !== toRight) {
-      this.nodesBetween(fromRight!, toRight!, (node) => {
-        if (!node.isRemoved) {
-          toBeRemoveds.push(node);
-        }
-      });
 
-      const isRangeOnSameBranch = toPos.node.isAncestorOf(fromPos.node);
-      for (const node of toBeRemoveds) {
-        node.remove(editedAt);
+    if (fromLeft !== toLeft) {
+      let fromChildIndex;
+      let parent;
 
-        if (node.isRemoved) {
-          this.removedNodeMap.set(node.pos.toIDString(), node);
-        }
+      if (fromLeft.parent === toLeft.parent) {
+        parent = fromLeft.parent!;
+        fromChildIndex = parent.allChildren.indexOf(fromLeft) + 1;
+      } else {
+        parent = fromLeft;
+        fromChildIndex = 0;
       }
 
-      // move the alive children of the removed element node
-      if (isRangeOnSameBranch) {
-        let removedElementNode: CRDTTreeNode | undefined;
-        if (fromPos.node.parent?.isRemoved) {
-          removedElementNode = fromPos.node.parent;
-        } else if (!fromPos.node.isText && fromPos.node.isRemoved) {
-          removedElementNode = fromPos.node;
+      const toChildIndex = parent.allChildren.indexOf(toLeft);
+
+      for (let i = fromChildIndex; i <= toChildIndex; i++) {
+        const node = parent.allChildren[i];
+
+        if (
+          node.getCreatedAt().getLamportAsString() ===
+            editedAt.getLamportAsString() &&
+          node.getCreatedAt().getActorID() !== editedAt.getActorID()
+        ) {
+          continue;
```

Comment:
Is this condition for dealing with concurrent editing situations? 
I guess it is not possible to determine whether operations are concurrent **using Lamport**.
Do we have to introduce a vector clock for this?

ref. https://cs.stackexchange.com/questions/101496/difference-between-lamport-timestamps-and-vector-clocks

Reply:
Yes it is, 
can you tell me why this is not possible to determine concurrency???